Add no-stdlib build for mlkem-native

willieyz opened the pull request:
- Resolves #963 
- This PR add an example demonstrating how to build mlkem-native without the standard library (`-nostdlib`)
  - Create an example folder named `multilevel_build_no_stdlib`
  - Add the `example_no_stdlib_config.h` reference from `test/custom_stdlib_config.h`,
     this config file provide custom implementations for `mlk_memcpy` and `mlk_memset`
  - Add the `-nostdlib` cflag during generating objects file.
  - Integrate this example to the `tests` script and `./Makefile`

willieyz committed:
Example: add the `multilevel_build_no_stdlib` example

- This commit add an example demonstrating how to build mlkem-native without the standard library (-nostdlib)
- Create an example folder named `multilevel_build_no_stdlib`
- Add the `example_no_stdlib_config.h` reference from `test/custom_stdlib_config.h`,
  this config provide custom implementations for mlk_memcpy and mlk_memset
- Add the `-nostdlib` cflag during generating objects file.
- Integrate this example to the `tests` script and ./Makefile
- Add this example to the `base.yml` for CI testing

Signed-off-by: willieyz <willie.zhao@chelpis.com>

willieyz commented on examples/multilevel_build_no_stdlib/README.md:
Yes, thanks for the reminder!
I have addressed this and re-pushed the changes,
including `examples/basic/mlkem_native/mlkem/README.md` and `examples/README.md`,
and also added them to `base.yml` for the CI tests.

hanno-becker commented on examples/README.md:
```diff
@@ -13,6 +13,12 @@ See [basic](basic) for a basic example of how to build a single instance of mlke
+## Multi-level build with no stdlib (C only)
+
+See [multilevel_build_no_stdlib](multilevel_build_no_stdlib) for an example of how to build one instance of mlkem-native per security level,
+with `-nostdlib` flag, the `mlk_memcpy` and `mlk_memset` are replaced by custom implementation.
+in such a way that level-independent code with no stadard libary is shared.
```
Please fix the grammar

hanno-becker commented on examples/multilevel_build_no_stdlib/Makefile:
```diff
@@ -0,0 +1,128 @@
+	-Wno-unknown-pragmas \
+	-Wno-unused-command-line-argument \
+	-fomit-frame-pointer \
+        -DMLK_CONFIG_NAMESPACE_PREFIX=mlkem \
```
nit: indentation

mkannwischer commented on examples/README.md:
```diff
@@ -13,6 +13,12 @@ See [basic](basic) for a basic example of how to build a single instance of mlke
+## Multi-level build without the standard library (C only)
+
+See multilevel_build_no_stdlib for an example of how to build one instance of mlkem-native per security level without the standard library.
+In this example, mlk_memcpy and mlk_memset are replaced with custom implementations through `custom_no_stdlib_config.h` and by using the -nostdlib flag,
+This approach allows level-independent code to be shared without relying on the standard C library.
```
```suggestion
In this example, `mlk_memcpy` and `mlk_memset` are replaced with custom implementations through [custom_no_stdlib_config.h](custom_no_stdlib_config.h).
```

mkannwischer commented on scripts/tests:
```diff
@@ -208,6 +208,7 @@ class TEST_TYPES(Enum):
     MONOLITHIC_BUILD_NATIVE = 14
     STACK = 15
     SIZE = 16
+    MULTILEVEL_BUILD_NO_STDLIB = 17
```
This conflicts with https://github.com/pq-code-package/mlkem-native/pull/1185.
Can you rebase this PR on top of #1185, please? We will likely merge that one first.

willieyz replied:
Rebased.
Thank you for taking the time to review!

mkannwischer reviewed (changes requested):
I am fine with the changes now (except for a small nit), but this PR conflicts with #1185. Please rebase on top of that one.

mkannwischer reviewed (changes requested):
Please rebase this on top of main.

mkannwischer commented on examples/multilevel_build_no_stdlib/README.md:
```diff
@@ -0,0 +1,18 @@
+# Multi-level build with no standard library
+
+This directory contains a minimal example for how to build mlkem-native with support for all 3 security levels
+MLKEM-512, MLKEM-768, and MLKEM-1024, and so that level-independent code is shared. In this example, only the C-backend
```
The parameter sets are called ML-KEM-512 etc. Can you please change that *everywhere* in mlkem-native?

mkannwischer reviewed (changes requested):
Thanks @willieyz. I am fine with the code changes now.
I found a small typo in the READMEs that was pre-existing before your PR, but maybe we can sort that out in a separate commit everywhere?

mkannwischer commented on examples/multilevel_build_no_stdlib/README.md:
```diff
@@ -0,0 +1,18 @@
+
+# Multi-level build with no standard library
+
+This directory contains a minimal example for how to build mlkem-native with support for all 3 security levels
```
This description is copied 1:1 from the multilevel build? Can you please add a sentence about nostd lib?

willieyz committed:
Rename all parameter sets: from `MLKEM-{512/768/1024}` to `ML-KEM-{512/768/1024}`
- This typo was pre-existing before this PR #1153, we sort this out with a separate commit

Signed-off-by: willieyz <willie.zhao@chelpis.com>

mkannwischer commented on examples/multilevel_build_no_stdlib/README.md:
```diff
@@ -0,0 +1,17 @@
+The library is built 3 times in different build directories `build/mlkem{512,768,1024}`. For the ML-KEM-512 build, we set
+`MLK_CONFIG_MULTILEVEL_WITH_SHARED` to force the inclusion of all level-independent code in the
+MLKEM512-build. For ML-KEM-768 and ML-KEM-1024, we set `MLK_CONFIG_MULTILEVEL_NO_SHARED` to not include any
+level-independent code, beside this, this example replace the `mlk_memcpy`, `mlk_memset` with the custom implementation memcpy and memset instead of using the stdlib by adding `example_no_stdlib_config.h`, and add the `-nostdlib` during objects file building,
```
```suggestion
level-independent code, beside this, this example replace the `mlk_memcpy`, `mlk_memset` with the custom implementation memcpy and memset instead of using the stdlib by adding `example_no_stdlib_config.h`, and add `-nostdlib` to the compilation flags,
```

mkannwischer reviewed (changes requested):
Small correction of the grammar. Then this is good to go. 
Thanks @willieyz!

mkannwischer reviewed (approved):
Thanks @willieyz. I'm happy with this PR now. 
@hanno-becker - what do you think? Note that the second commit (renaming) is somewhat unrelated and requested by me.

mkannwischer reviewed (changes requested):
This PR predates https://github.com/pq-code-package/mlkem-native/pull/1227 and _did not_ catch that there actually was a dependency on the stdlib (memcmp) - so it clearly does not achieve what https://github.com/pq-code-package/mlkem-native/issues/963 is asking for. 
I will change this back to draft and we should probably close it unless we can think of a way to implement it properly.

mkannwischer commented:
Closing until we can think of a way to implement it properly.
